src/writer.rs

```diff
@@ -0,0 +1,200 @@
+//! Writer for writer functions
+
+use bitvec::bitvec;
+use bitvec::{field::BitField, prelude::*};
+use no_std_io::io::Write;
+
+#[cfg(feature = "logging")]
+use log;
+
+use crate::DekuError;
+
+const fn bits_of<T>() -> usize {
+    core::mem::size_of::<T>().saturating_mul(<u8>::BITS as usize)
+}
+
+/// Container to use with `to_writer`
+pub struct Writer<W: Write> {
+    pub(crate) inner: W,
+    /// Leftover bits
+    pub leftover: BitVec<u8, Msb0>,
+    /// Total bits written
+    pub bits_written: usize,
+}
+
+impl<W: Write> Writer<W> {
+    /// Create a new `Writer`
+    #[inline]
+    pub fn new(inner: W) -> Self {
+        Self {
+            inner,
+            leftover: BitVec::new(),
+            bits_written: 0,
+        }
+    }
+
+    /// Return the unused bits
+    #[inline]
+    pub fn rest(&mut self) -> alloc::vec::Vec<bool> {
+        self.leftover.iter().by_vals().collect()
+    }
+
+    /// Write all `bits` to `Writer` buffer if bits can fit into a byte buffer.
+    ///
+    /// Any leftover bits will be written before `bits`, and non-written bits will
+    /// be stored in `self.leftover`.
+    #[inline]
+    pub fn write_bits(&mut self, bits: &BitSlice<u8, Msb0>) -> Result<(), DekuError> {
+        #[cfg(feature = "logging")]
+        log::trace!("attempting {} bits", bits.len());
+
+        // quick return if we can't write to the bytes buffer
+        if (self.leftover.len() + bits.len()) < 8 {
+            self.leftover.extend_from_bitslice(bits);
+            return Ok(());
+        }
+
+        // pre-pend the previous attempt to write if needed
+        let mut bits = if self.leftover.is_empty() {
+            bits
+        } else {
+            #[cfg(feature = "logging")]
+            log::trace!("pre-pending {} bits", self.leftover.len());
+
+            self.leftover.extend_from_bitslice(bits);
+            &mut self.leftover
+        };
+
+        // one shot impl of BitSlice::read(no read_exact), but for no_std
+        let mut buf = alloc::vec![0x00; bits.len() / 8];
+        let mut count = 0;
+        bits.chunks_exact(bits_of::<u8>())
+            .zip(buf.iter_mut())
+            .for_each(|(byte, slot)| {
+                *slot = byte.load_be();
+                count += 1;
+            });
+
+        // SAFETY: This does noto have a safety comment in bitvec. But this is safe
+        // because of `count` here will always still be within the bounds
+        // of `bits`
+        bits = unsafe { bits.get_unchecked(count * bits_of::<u8>()..) };
+
+        // TODO: with_capacity?
```

Comment:
TODO, clarify or address

Do you mean we could infer `self.leftover` capacity at initialization?

Reply:
Changed, not applicable.